File: src/deadlock_build_sync/offline/discovery_admission.py
"""Apply frozen core, order, mechanics, and pool gates to validation data."""

from scipy.stats import norm

from .discovery_data import HeroData
from .discovery_orders import core_times, order_evidence
from .discovery_quality import evaluate_core, rejection_reasons
from .discovery_types import Nomination


def admit_core(
    values: HeroData, row: Nomination, family: int, frozen_hash: str
) -> Nomination:
    validation = evaluate_core(values, tuple(row["items"]), "validation")
    limitations = [
        f"selection: {reason}" for reason in rejection_reasons(row["selection"])
    ]
    limitations.extend(
        f"validation: {reason}" for reason in rejection_reasons(validation, family)
    )
    reasons = list(row["selection_rejections"])
    adjusted = validation["adjusted"]
    validation["adjusted_lower_family"] = (
        adjusted["difference"]
        - float(norm.isf(0.025 / family)) * adjusted["standard_error"]
        if adjusted["difference"] is not None
        else None
    )
    ordered = order_evidence(
        core_times(values, row["items"], "validation"),
        row["items"],
        row["path"]["order"],
    )
    if not ordered["passes"]:
        limitations.append("validation: frozen order lacks support")
    if not row["path"]["admitted_before_validation"]:
        reasons.append("Frozen order lacks discovery or selection support")
    if not row["tactics"]["supported_focus"]:
        limitations.append(
            row["tactics"]["reason"] or "Mechanic text match is unavailable"
        )
    if not row["guide"]["ready"]:
        reasons.append(row["guide"]["reason"] or "Unsupported guide")
    return {
        **row,
        "evidence_status": "observed" if limitations else "outcome_supported",
        "evidence_limitations": limitations,
        "validation": validation,
        "order_validation": ordered,
        "hypotheses": family,
        "rejections": reasons,
        "frozen_sha256": frozen_hash,
    }


def discovery_record(row: Nomination) -> dict[str, object]:
    record = {
        key: value
        for key, value in row.items()
        if key not in {"guide", "automatic_choices", "branch_candidates"}
    }
    return {
        **record,
        "method": "eclat_leiden_pairwise",
        "frozen_guide": row["guide"],
        "test_evaluated": False,
    }

File: crates/deadlock-analysis/src/discovery_admission.rs
use deadlock_data::{Result, count_as_f64, normal_quantile, real};
use deadlock_guides::OutcomeEvidence;
use serde_json::{Value, json};

use crate::core_outcomes::evaluate_core;
use crate::discovery_data::DiscoveryData;
use crate::discovery_models::{Nomination, item_ids};
use crate::purchase_orders::order_evidence;

pub fn admit_nomination(
    data: &DiscoveryData,
    nominee: &Nomination,
    family: u64,
    frozen_hash: &str,
) -> Result<Value> {
    let mut validation = evaluate_core(data, &nominee.candidate.items, "validation")?;
    let selection = OutcomeEvidence::from_document(&nominee.candidate.selection)?;
    let mut limitations = selection
        .limitations(None)?
        .into_iter()
        .map(|reason| format!("selection: {reason}"))
        .collect::<Vec<_>>();
    limitations.extend(
        OutcomeEvidence::from_document(&validation)?
            .limitations(Some(family))?
            .into_iter()
            .map(|reason| format!("validation: {reason}")),
    );
    validation["adjusted_lower_family"] = if validation["adjusted"]["difference"].is_null() {
        Value::Null
    } else {
        (normal_quantile(1.0 - 0.025 / count_as_f64(family.max(1))?)?)
            .mul_add(
                -(real(&validation["adjusted"], "standard_error")?),
                real(&validation["adjusted"], "difference")?,
            )
            .into()
    };
    let ordered = order_evidence(
        data,
        &nominee.candidate.items,
        &item_ids(&nominee.path["order"])?,
        "validation",
    )?;
    if ordered["passes"] != true {
        limitations.push("validation: Frozen order lacks support".into());
    }
    if nominee.tactics["supported_focus"] != true {
        limitations.push(
            nominee.tactics["reason"]
                .as_str()
                .unwrap_or("Mechanic text comparison is unavailable")
                .into(),
        );
    }
    let mut reasons = nominee.candidate.selection_rejections.clone();
    if nominee.path["admitted_before_validation"] != true {
        reasons.push("Frozen order lacks discovery or selection support".into());
    }
    if !nominee.guide.ready {
        reasons.push(
            nominee
                .guide
                .reason
                .clone()
                .unwrap_or_else(|| "Unsupported guide".into()),
        );
    }
    let mut result = serde_json::to_value(nominee)?;
    result["evidence_status"] = if limitations.is_empty() {
        "outcome_supported"
    } else {
        "observed"
    }
    .into();
    result["evidence_limitations"] = json!(limitations);
    result["validation"] = validation;
    result["order_validation"] = ordered;
    result["hypotheses"] = family.into();
    result["rejections"] = json!(reasons);
    result["frozen_sha256"] = frozen_hash.into();
    Ok(result)
}

pub fn discovery_record(row: &Value) -> Result<Value> {
    let mut result = deadlock_data::object(row)?.clone();
    for key in ["guide", "automatic_choices", "branch_candidates"] {
        result.remove(key);
    }
    result.insert("method".into(), "eclat_leiden_pairwise".into());
    result.insert("frozen_guide".into(), row["guide"].clone());
    result.insert("test_evaluated".into(), false.into());
    Ok(result.into())
}
